lib/system/web/src/keyboard_engine.rs

```diff
@@ -0,0 +1,122 @@
+use wasm_bindgen::JsCast;
+use wasm_bindgen::prelude::*;
+
+use crate::*;
+
+type Bindings = HashMap<BTreeSet<u32>, Vec<Box<dyn FnMut()>>>;
+
+/// Keyboard Engine that handles user keyboard combinations
+///
+/// Keyboard Engine stores bindings as HashMap:
+/// {
+///     [keyA, keyB]:          [Callback1, Callback2],
+///     [keyA, keyB, keyCtrl]: [Callback1],
+/// }
+/// when the combination is pressed related callbacks are invoked
+pub struct KeyboardEngine {
+    pub bindings: Rc<RefCell<Bindings>>,
+    pub state:    Rc<RefCell<BTreeSet<u32>>>,
+    pub target:   web_sys::EventTarget
+}
+
+impl KeyboardEngine {
+    pub fn new(target: &web_sys::EventTarget) -> Result<Self> {
+        let state = Rc::new(RefCell::new(BTreeSet::new()));
+        let bindings = Rc::new(RefCell::new(HashMap::new()));
+
+        KeyboardEngine::init(target, Rc::clone(&bindings), Rc::clone(&state));
+
+        Ok(Self {
+            bindings,
+            state,
+            target: target.clone()
+        })
+    }
+
+    pub fn from_tag_name(html_tag: &str) -> Result<Self> {
+        let element = document()?
+            .get_elements_by_tag_name(html_tag)
+            .item(0).ok_or_else(|| Error::missing("there is no such element"));
+
+        KeyboardEngine::new(&element.unwrap())
+    }
+
+    fn init (
+        target:   &web_sys::EventTarget,
+        bindings: Rc<RefCell<Bindings>>,
+        state:    Rc<RefCell<BTreeSet<u32>>>
+    ) {
+        {
+            let state = Rc::clone(&state);
+            let bindings = Rc::clone(&bindings);
+            let keydown_callback = Box::new(move |event: web_sys::KeyboardEvent| {
+                if event.repeat() {
+                    return;
+                }
+
+                let key_code = event.key_code();
+                state.borrow_mut().insert(key_code);
+
+                let key = state.borrow();
+                if let Some(vec) = bindings.borrow_mut().get_mut(&key) {
+                    vec.iter_mut().for_each(|func| func());
+                }
+            });
+            let keydown_callback = Closure::wrap(
+                keydown_callback as Box<dyn FnMut(_)>);
+            target.add_event_listener_with_callback(
+                "keydown", keydown_callback.as_ref().unchecked_ref())
+                .unwrap();
+            keydown_callback.forget();
+        }
+
+        {
+            let keyup_callback = Box::new(move |event: KeyboardEvent| {
+                let key_code = event.key_code();
+                state.borrow_mut().remove(&key_code);
+
+                let key = state.borrow();
+                if let Some(vec) = bindings.borrow_mut().get_mut(&key) {
+                    vec.iter_mut().for_each(|func| func());
+                }
+            });
+            let keyup_callback = Closure::wrap(
+                keyup_callback as Box<dyn FnMut(_)>);
+            target.add_event_listener_with_callback(
+                "keyup", keyup_callback.as_ref().unchecked_ref())
```

Comment:
This code is definitely too long and contains some strange sections inside. We try to write a code where each function has max of 10-15 lines. If it exceeds that limit, it should almost for sure be refactred to several functions. Basically reading the code line by line should be like reading a book - each line should contain a function call telling what is happening there. Here we got something much more complex, which is hard to read and hard to track.